File: leetcode/09-Search-in-Rotated-Sorted-Array/python/main.py
import math


class Solution:
    def search(self, nums, target: int) -> int:
        head_index = 0
        tail_index = len(nums) - 1

        result = -1

        while head_index <= tail_index:
            mid_index = math.floor((head_index + tail_index) / 2)

            if nums[mid_index] == target:
                result = mid_index

            if nums[head_index] <= nums[mid_index]:
                # 左側がソートされている場合（右側のどこかで境界がある）
                if nums[head_index] <= target and target < nums[mid_index]:
                    # 左側に存在しているので次のループは左へ
                    tail_index = mid_index - 1
                else:
                    # 左側には存在しないので次のループは右へ
                    head_index = mid_index + 1
            elif nums[mid_index] <= nums[tail_index]:
                # 右側がソートされている場合（左側のどこかで境界がある）
                if nums[mid_index] < target and target <= nums[tail_index]:
                    # 右側に存在しているので次のループは右へ
                    head_index = mid_index + 1
                else:
                    # 右側には存在していないので次のループは左へ
                    tail_index = mid_index - 1

        return result

File: leetcode/09-Search-in-Rotated-Sorted-Array/go/main.go
package main

import "fmt"

func search(nums []int, target int) int {
	left, right := 0, len(nums)-1

	result := -1

	fmt.Println()
	for left <= right {
		mid := (left + right) / 2 // int型なので切り捨て
		fmt.Printf("Start left: %v, mid: %v, right: %v\n", left, mid, right)

		if nums[mid] == target {
			result = mid
			break
		}

		if nums[left] <= nums[mid] {
			// 境界が右側にある場合
			if nums[left] <= target && target < nums[mid] {
				// 左側に存在するので次のループは左側
				right = mid - 1
			} else {
				// 右側に存在するので次のループは右側
				left = mid + 1
			}
		} else {
			// 境界が左側にある場合
			if nums[mid] < target && target <= nums[right] {
				// 右側に存在するので次のループは右側
				left = mid + 1
			} else {
				// 左側に存在するので次のループは左側
				right = mid - 1
			}
		}
		fmt.Printf("End left: %v, mid: %v, right: %v\n", left, mid, right)

	}

	return result
}
